Nwa50ax

telent opened the pull request:
Hi @RaitoBezarius  and thanks for the new device. I opened this PR on your branch so I can make notes on which bits I have taken already/will take/have fixed at source. If you want to rebase against a more recent main version you should already find that some of this goes away, but please don't feel you're required to: I'm happy to work with what's here and preserve attribution.

RaitoBezarius committed:
devices/zyxel-nwa50ax: init

RaitoBezarius committed:
builders/uimage: support aligning the FIT

This is necessary when writing to a MTD partition with a certain erasesize.

RaitoBezarius committed:
modules/outputs/ubivolume: introduce ubinization

It creates an UBI image based on an UBI volume configuration.

For now, it creates only an empty rootfs.

RaitoBezarius committed:
modules/ubifs: revamp to offer directly access to the UBIfs partition

Adds the LEB and PEB option and let the user remove the boot image in case
where U-Boot does not support UBI boot.

RaitoBezarius committed:
zyxel-bootconfig: init at no version

This tool is useful for manipulating the A/B boot status of the image.

telent commented on default.nix:
```diff
@@ -12,7 +12,7 @@ let
-        "python-2.7.18.6"       # kernel backports needs python <3
+        "python-2.7.18.7"       # kernel backports needs python <3
       ];
     };
```
cherry-picked this

telent commented on default.nix:
```diff
@@ -22,6 +22,7 @@ let
       ./modules/hardware.nix
       ./modules/base.nix
+      ./modules/standard.nix
       ./modules/busybox.nix
```
I *think* I've made this go away, in favour of moving outputs into the devices/xxx/default.nix modules

telent commented on devices/zyxel-nwa50ax/default.nix:
```diff
@@ -0,0 +1,300 @@
+    with restricted boot commands. Fortunately, OpenWrt folks figured out trivial command injections,
+    so you can use most of the OpenWrt commands without trouble by just command injecting
+    atns, atna or atnf, e.g. atns "; $real_command".
+
```
fwiw I have the germ of an idea for adding per-device customisation in the boot.scr/flash.scr  so that this could be taken care of (it's looking like the Turris Omnia also has its own requirements for tftpboot) - I just haven't got that far yet

telent commented on devices/zyxel-nwa50ax/default.nix:
```diff
@@ -0,0 +1,300 @@
+
+    OpenWrt web page: https://openwrt.org/inbox/toh/zyxel/nwa50ax
+    OpenWrt tech data: https://openwrt.org/toh/hwdata/zyxel/zyxel_nwa50ax
+
```
Nice writeup, btw :+1:

telent commented on devices/zyxel-nwa50ax/default.nix:
```diff
@@ -0,0 +1,300 @@
+        flash = {
+          address = lim.parseInt "0xbc050000";
+          size = lim.parseInt "0xfb0000";
+          eraseBlockSize = 65536;
```
this  eraseBlockSize is inconsistent with `ubi.physicalEraseBlockSize` but I'm guessing it's probably just not used?

RaitoBezarius replied:
Yep, I will get rid of that block.

telent commented on devices/zyxel-nwa50ax/default.nix:
```diff
@@ -0,0 +1,300 @@
+          # 20MB seems to give enough room to uncompress the kernel
+          # without anything getting trodden on. 10MB was too small
+          loadAddress = lim.parseInt "0x2000000";
+          useMemmap = true;
```
https://gti.telent.net/dan/liminix/commit/231c2cef0323d23ea61b9780442b19ba140a985a I think I fixed the bug that was causing it to need this

telent commented on modules/initramfs.nix:
```diff
@@ -53,6 +53,11 @@ in
+          nod /dev/mtdblock8 0600 0 0 b 31 5
+          nod /dev/mtdblock9 0600 0 0 b 31 5
+          nod /dev/mtdblock10 0600 0 0 b 31 5
           dir /target 0755 0 0
```
am now using devtmpfs in initramfs (because of /dev/vda which doesn't have a static major number), so this is not needed any more

telent commented on pkgs/kernel/default.nix:
```diff
@@ -90,7 +90,7 @@ stdenv.mkDerivation rec {
   buildPhase = ''
-    make vmlinux modules_prepare
+    make vmlinux modules_prepare -j$NIX_BUILD_CORES
   '';
```
cherry-picked this

telent reviewed (commented):
Hi, and thanks for the new device.  I opened a PR on your branch so I can make notes on which bits I have taken already/will take/have fixed at source.  If you want to rebase against a more recent `main` version you should already find that some of this goes away, but  please don't feel you're required to: I'm happy to work with what's here and preserve attribution.

RaitoBezarius committed:
pkgs/kernel/uimage: introduce `commandLineDtbNode`

Certain devices like the Zyxel NWA50AX will pass information on the command-line
to explain what is the current image (`bootImage=1` vs. `bootImage=0`).

Unfortunately, if we set the `chosen/bootargs` node, this will be overridden forcibly
by U-Boot.

To avoid this problem, it's easier to simply just use another DTB node like `bootargs-override` which
is what OpenWRT does [1].

[1]: https://git.openwrt.org/?p=openwrt/openwrt.git;a=blob;f=target/linux/ramips/patches-5.15/314-MIPS-add-bootargs-override-property.patch;h=e7dca7af886e8c0b69ba2b23f5855ddfeeb0d4a1;hb=refs/heads/main

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius committed:
modules/outputs: expose `commandLineDtbNode` option

We allow `bootargs` and `bootargs-override` for now only.

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius committed:
modules/outputs/ubifs: expose `rootubifs` rather than `rootfs`

I believe there should be another module exposing `rootubifs` as `rootfs`
or let any other module just subsume that component like `zyxel-nwa-fit` output.

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius committed:
modules/outputs/zyxel-nwa-fit: init

Zyxel "firmware" format is just… a FIT with some metadata on the models.

This FIT is like this:

--------------------------
    uImage FIT header
--------------------------
    Linux kernel
--------------------------
    FDT DTB
--------------------------
    Padding so that
    this makes
    8192kb [1]
--------------------------
    UBI volume
    as a root filesystem
--------------------------

We just reproduce this in a very brutal and naive way.
In the future, this seems worth to generalize and modularize this idea
so that zyxel-nwa-fit is just an instance of a more general output.

[1]: https://git.openwrt.org/?p=openwrt/openwrt.git;a=blob;f=target/linux/ramips/image/mt7621.mk;h=ab1b829ba0086cb9fc9ca8cbbf3cbc14735034d6;hb=refs/heads/main#l3097

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius committed:
examples/nwa50ax-ap: init

This is a quite comprehensive example using maximally the hardware
available to reach nice performance.

In the future, I will even add RADIUS examples.

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

telent commented on devices/zyxel-nwa50ax/default.nix:
```diff
@@ -0,0 +1,310 @@
+          freeSpaceBytes = 5 * 1024 * 1024;
+          # 20MB seems to give enough room to uncompress the kernel
+          # without anything getting trodden on. 10MB was too small
+          loadAddress = lim.parseInt "0x2000000";
```
nit: don't think the comment matches the code here?

RaitoBezarius replied:
Yes, that's true.

RaitoBezarius replied:
Done.

telent commented on modules/hardware.nix:
```diff
@@ -67,6 +67,7 @@ in {
       };
       loadAddress = mkOption { type = types.ints.unsigned; default = null; };
       entryPoint = mkOption { type = types.ints.unsigned; };
+      alignment = mkOption { type = types.ints.unsigned; default = null; };
```
can you put a description on this to say what it's the alignment of?

RaitoBezarius replied:
Yep.

RaitoBezarius replied:
Done.

telent commented on modules/outputs.nix:
```diff
@@ -111,7 +111,8 @@ in
-        inherit (config.hardware) loadAddress entryPoint;
+        inherit (config.boot) commandLineDtbNode;
+        inherit (config.hardware) loadAddress entryPoint alignment;
         inherit (config.boot) imageFormat;
```
comment: At some future point we'll have to tidy up the 'boot' vs 'hardware' config space because it's a bit of a mess right now.

telent commented on pkgs/kernel/uimage.nix:
```diff
@@ -15,10 +15,12 @@ let
 in {
   kernel
 , commandLine
+, commandLineDtbNode ? "bootargs"
```
:+1:

telent commented on pkgs/min-copy-closure/liminix-rebuild.sh:
```diff
@@ -20,7 +20,7 @@ if toplevel=$(nix-build "$@" -A outputs.systemConfiguration --no-out-link); then
     min-copy-closure $target_host $toplevel
     $ssh_command $target_host $toplevel/bin/install
-    $ssh_command  $target_host "sync; source /etc/profile; reboot"
+    $ssh_command  $target_host "sync; source /etc/profile; reboot -f"
```
This is a hack but I'm not going to be  in a position to fix it properly  for a few weeks, so it only automates what i'm doing by hand anyway - fair enough.  https://gti.telent.net/dan/liminix/issues/9

telent reviewed (commented):
Looks good! Some initial nitpicks. I'll have a proper read through and get my head around the ubifs stuff as soon as I've had a coffee

telent commented on devices/zyxel-nwa50ax/default.nix:
```diff
@@ -0,0 +1,310 @@
+      mac80211 = pkgs.mac80211.override {
+        drivers = [ "mt7915e" ];
+        klibBuild = config.system.outputs.kernel.modulesupport;
+      };
```
Note: The Future here is to switch from  pkgs.mac80211 (which uses backports) to pkgs.kmodloader, which uses the modules in the current kernel.See  https://gti.telent.net/dan/liminix/commit/65479e206b62a9b269d3825f8cc5e9d660d583ec 

Caveat: I haven't yet tested it on a mt7915e and it may be that by doing this we've lost necessary openwrt patches. See  e.g. https://gti.telent.net/dan/liminix/commit/ca64e9035e0e9792059faa5fcaaee2b959bed458 for an example of restoring them

RaitoBezarius committed:
modules/busybox: add `dhcprelay`

This enables to run a DHCP relay from multiple interfaces.

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius commented:
@telent PTAL again, I rebased some stuff.

RaitoBezarius committed:
modules/zyxel-dual-image: init

This adds a simple boot blessing module, to be used, with the Zyxel NWA50AX.

There's a lot of elephant in the rooms: how do you upgrade kernel, etc.

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius committed:
devices/zyxel-nwa50ax: use our own more advanced DTB

OpenWRT had a DTB for the NWA50AX LEDs that I didn't pick up.

Anyway, we need to include our own special DTB for the NWA platform in general
to support A/B operations, because OpenWRT original one just mark everything else read-only.

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius committed:
devices/zyxel-nwa50ax: ensure bridge is always available

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius committed:
devices/zyxel-nwa50ax: cleanup of `flash` attribute and `rootDevice`

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius committed:
examples/nwa50ax-ap: ensure `mtdutils` is available for further flashing

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius committed:
examples/nwa50ax-ap: support bridge between lan and ethernet

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius committed:
devices/zyxel-nwa50ax: expose primary and secondary images

To support A/B a bit better.

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius committed:
devices/zyxel-nwa50ax: upgrade MT7915 firmware from OpenWRT repository

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius committed:
devices/zyxel-nwa50ax: complete documentation

It covers everything I know more or less.

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius committed:
devices/zyxel-nwa50ax: developer todo

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius committed:
devices/zyxel-nwa50ax: make `zyxel-bootconfig` executable

Otherwise, it doesn't work well…

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius committed:
devices/zyxel-nwa50ax: fix dual image mgmt after DTB expansion

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

telent commented on devices/zyxel-nwa50ax/default.nix:
```diff
@@ -0,0 +1,367 @@
+
+          PRINTK_TIME = "y";
+        } // lib.optionalAttrs (config.system.service ? vlan) {
+          SWCONFIG = "y";
```
don't think you need SWCONFIG if you're using DSA, I'd assume it's one or the other

telent commented on modules/hardware.nix:
```diff
@@ -67,6 +67,7 @@ in {
       };
       loadAddress = mkOption { type = types.ints.unsigned; default = null; };
       entryPoint = mkOption { type = types.ints.unsigned; };
+      alignment = mkOption { type = types.ints.unsigned; default = null; description = "Alignment passed to `mkimage` for FIT"; };
```
:+1:

flokli commented on modules/outputs/ubifs.nix:
```diff
@@ -12,9 +12,16 @@ in
   options.hardware.ubi = {
     minIOSize = mkOption { type = types.str; };
-    eraseBlockSize = mkOption { type = types.str; }; # LEB
+    logicalEraseBlockSize = mkOption { type = types.str; }; # LEB
```
This is missing some pointers on how to obtain this information - the `ubinfo` command only shows logical erase block size.
